src/DataAPI.jl

```diff
@@ -287,4 +287,131 @@ using a `sink` function to materialize the table.
+
+Functions adding methods to `metadata` should define them only with
+`key::AbstractString` signature. Passing `key` that is not a string to
+`metadata` must throw `MethodError`.
+
+If `style=true` return a tuple of metadata value and metadata style. Metadata
+style is an additional information about the kind of metadata that is stored
+for the `key`.
+
+$STYLE_INFO
+"""
+metadata(::T, ::AbstractString; style::Bool=false) where {T} =
+    throw(ArgumentError("Objects of type $T do not support getting metadata"))
+
+"""
+    metadatakeys(x)
+
+Return an iterator of metadata keys for which `metadata(x, key)` returns a
+metadata value. If `x` does not support metadata return `()`.
+"""
+metadatakeys(::Any) = ()
+
+"""
+    metadata!(x, key::AbstractString, value; style)
+
+Set metadata for object `x` for key `key` to have value `value` and style `style`
+and return `x`.
+If `x` does not support setting metadata throw `ArgumentError`.
+
+Functions adding methods to `metadata!` should define them only with
+`key::AbstractString` signature. Passing `key` that is not a string to
+`metadata` must throw `MethodError`.
+
+$STYLE_INFO
+"""
+metadata!(::T, ::AbstractString, ::Any; style) where {T} =
+    throw(ArgumentError("Objects of type $T do not support setting metadata"))
+
+"""
+    colmetadata(x, col, key::AbstractString; style::Bool=false)
+
+Return metadata value associated with table `x` for column `col` and key `key`.
+If `x` does not support metadata for column `col` throw `ArgumentError`. If `x`
+supports metadata, but does not have a mapping for column `col` for `key` throw
+`KeyError`.
+
+Functions adding methods to `colmetadata` should define them only with
+`key::AbstractString` signature. Passing `key` that is not a string to
+`metadata` must throw `MethodError`.
+
+$COL_INFO
+
+If `style=true` return a tuple of metadata value and metadata style. Metadata
+style is an additional information about the kind of metadata that is stored for
+the `key`.
+
+$STYLE_INFO
+"""
+colmetadata(::T, ::Int, ::AbstractString; style::Bool=false) where {T} =
+    throw(ArgumentError("Objects of type $T do not support getting column metadata"))
+colmetadata(::T, ::Symbol, ::AbstractString; style::Bool=false) where {T} =
+    throw(ArgumentError("Objects of type $T do not support getting column metadata"))
+
+"""
+    colmetadatakeys(x, [col])
+
+If `col` is passed return an iterator of metadata keys for which
+`metadata(x, col, key)` returns a metadata value.
+If `x` does not support metadata for column `col` return `()`.
+
+`col` must be a type that is supported by table `x`. Following the Tables.jl
+contract `Symbol` and `Int` must be supported. However, tables that allow other
+column indexing (e.g., using strings or integers other than `Int`) are allowed
+to accept such types of `col` argument. Passing `col` that is not a column
+of `x` can either throw an error or return `()` (this duality is allowed as
+some Tables.jl do not have schema).
```

Comment:
@nalimilan - this is the only case left that is problematic. The issue is that we cannot always check if `col` is a column in `x`.

Reply:
Ah right. In theory we could define the fallback only in Tables.jl so that it throws an error when the table has a schema and the column doesn't exist, but maybe this isn't worth the additional complexity. As long as common table types such as DataFrames throw an error in that case, users will likely not rely on the fact that some types accept non-existent columns.